routes/bridges/gassistant/index.js

```diff
@@ -0,0 +1,235 @@
+        this._requestSignin = false;
+    }
+
+    send(text, icon) {
+        if (typeof this._buffer[this._buffer.length - 1] === 'string')
+            // If there is already a text reply immediately before, we merge text replies
+            // because Google Assistant limits at most 2 chat bubbles per turn
+            this._buffer[this._buffer.length - 1] += '\n' + text;
+        else
+            this._buffer.push(text);
+    }
+
+    sendPicture(url, icon) {
+        if (typeof this._buffer[this._buffer.length - 1] !== 'string')
+            // If there is no text reply immediately before, we add the URL
+            // because Google Assistant requires a chat bubble to accompany an Image
+            this._buffer.push(url);
+        this._buffer.push(new Image({
+            url: url,
+            alt: url,
+        }));
+    }
+
+    sendRDL(rdl, icon) {
+        this._buffer.push(new BasicCard({
+            title: rdl.displayTitle,
+            text: rdl.displayText,
+            buttons: new Button({
+                title: rdl.displayTitle,
+                url: rdl.webCallback,
+            }),
+            image: new Image({
+                url: rdl.pictureUrl,
+                alt: rdl.pictureUrl
+            }),
+            display: 'CROPPED',
+        }));
+    }
+
+    sendChoice(idx, what, title, text) {
+        let suggestions = [];
+        // Filter out buttons more than 25 characters long
+        // since Google Assistant has a cap of 25 characters
+        if (title.length <= 25)
+            suggestions.push(title.substring(0, 25));
+        else
+            console.log(`${title} exceeds max length of 25 characters`);
+        if (suggestions.length)
+            this._buffer.push(new Suggestions(suggestions));
+    }
+
+    sendButton(title, json) {
+        let suggestions = [];
+        // Filter out buttons more than 25 characters long
+        // since Google Assistant has a cap of 25 characters
+        if (title.length <= 25)
+            suggestions.push(title.substring(0, 25));
+        else
+            console.log(`${title} exceeds max length of 25 characters`);
+        if (suggestions.length)
+            this._buffer.push(new Suggestions(suggestions));
+    }
+
+    sendLink(title, url) {
+        if (url === '/user/register') {
+            this._requestSignin = true;
+        } else {
+            this._buffer.push(new Button({
+                title: title,
+                url: url,
+            }));
+        }
+    }
+
+    sendResult(message, icon) {
+        if (typeof this._buffer[this._buffer.length - 1] === 'string')
+            // If there is already a text reply immediately before, we merge text replies
+            // because Google Assistant limits at most 2 chat bubbles per turn
+            this._buffer[this._buffer.length - 1] += '\n' + message.toLocaleString(this._locale);
+        else
+            this._buffer.push(message.toLocaleString(this._locale));
```

Comment:
It would be nice to support some of the extended result types before falling back to text. For example, it looks like the [Media formatted object](https://github.com/stanford-oval/thingtalk/blob/master/lib/runtime/format_objects.js#L304) (type="media" in ThingTalk) can be supported: https://developers.google.com/assistant/conversational/rich-responses#media_responses

Reply:
Okay. But seems like Google Assistant only supports audio for media response. Is there a good way to check whether the message is video or audio? Or should I just do something like `message.url.endsWith("mp4")`?

Reply:
Well in that case let's not worry for now, we don't have any device that uses that yet (It was introduced to play youtube videos)